File: 503_NextGreaterElement_II.py
##59.42%
class Solution(object):
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        if not nums:
            return nums
        res = [-1] * len(nums)
        stack = []
        for i, num in enumerate(nums):
            while stack and num > nums[stack[-1]]:
                res[stack[-1]] = num
                stack.pop()
            stack.append(i)
        start = 0
        while stack and start < stack[-1]:
            if nums[start] > nums[stack[-1]]:
                res[stack[-1]] = nums[start]
                stack.pop()
            else:
                start += 1
        return res

##A more concise method		
def nextGreaterElements(self, nums):
        stack, res = [], [-1] * len(nums)
        for i in range(len(nums)) * 2:
            while stack and (nums[stack[-1]] < nums[i]):
                res[stack.pop()] = nums[i]
            stack.append(i)
        return res

File: 503_NextGreaterElement_II.cpp
//52.25%
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        if(nums.empty())
			return nums;
		
		int size(nums.size());
		vector<int> res(size, -1);
		stack<int> positions;
		for(int i = 0; i < nums.size(); ++i){
			while(!positions.empty() && nums[positions.top()] < nums[i]){
				res[positions.top()] = nums[i];
				positions.pop();
			}
			positions.push(i);
		}
		
		int start(0);
		while(!positions.empty() && start < positions.top()){
			if(nums[start] > nums[positions.top()]){
				res[positions.top()] = nums[start];
				positions.pop();
			}else
				++start;
		}
		return res;
    }
};

//This way is more concise
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        vector<int> next(n, -1);
        stack<int> s; // index stack
        for (int i = 0; i < n * 2; i++) {
            int num = nums[i % n]; 
            while (!s.empty() && nums[s.top()] < num) {
                next[s.top()] = num;
                s.pop();
            }
            if (i < n) s.push(i);
        }   
        return next;
    }